RevisionName templating and substitution.

muscionig opened the pull request:
<!-- Thanks for sending a pull request! -->

<!-- 
Are you using Knative? If you do, we would love to know!
https://github.com/knative/community/issues/new?template=ADOPTERS.yaml&title=%5BADOPTERS%5D%3A+%24%7BCOMPANY+NAME+HERE%7D
-->
fixes #77 

# Changes

This PR adds support for replacing the `revisionName` in the `PrometheusMetric` field. I discussed this with @skonto in the slack channel and it seemed a necessary feature to separate the scaling of different `revisions`.

## Minor Changes
<!-- 
Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! 

- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic
-->

- :gift: adds a templating feature that allows to substitute the `revisionName` under the `PrometheusQuery` annotation. The feature uses Golang templates.
- :gift: adds `unit` tests for the template substitution.
- :gift: adds `e2e` tests: added a method to setup a service with the templated `prometheusQuery` with 2 `revisions`, and a test to check scaling status of both while generating continuous traffic.

Usage:
```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: metrics-test
spec:
  template:
    metadata:
      labels:
        app: metrics-test
      annotations:
        ...
        autoscaling.knative.dev/metric: "http_requests_total"
        autoscaling.knative.dev/prometheus-query: "sum(rate(http_requests_total{pod=~\"{{ .revisionName }}.*\"}[1m]))"
```

I tested this with `Istio` and `kourier`.

Output of the `controller`:
```
### service with no substitution
autoscaler-keda 2024/11/20 06:42:59 PrometheusQuery rendered: sum(rate(http_requests_total{}[1m]))                                    
autoscaler-keda {"severity":"INFO","timestamp":"2024-11-20T06:42:59.698756553Z","logger":"hpaautoscaler","caller":"controller/controller.go:548","message":"Reconcile succeeded","commit":"6f54184-dirty","knative.dev/controller":"knative.dev.autoscaler-keda.pkg.reconciler.autoscaling.hpa.Reconciler","knative.dev/kind":"autoscaling.internal.knative.dev.PodAutoscaler","knative.dev/traceid":"12a29044-6fe7-442b-a31c-7913633203cf","knative.dev/key":"metrics-test-istio/metrics-test-istio-00001","duration":"10.55125ms"}


### service with substitution
autoscaler-keda 2024/11/20 06:43:12 PrometheusQuery rendered: sum(rate(http_requests_total{pod=~"metrics-test-00001.*"}[1m]))
autoscaler-keda {"severity":"INFO","timestamp":"2024-11-20T06:43:12.122399087Z","logger":"hpaautoscaler","caller":"hpa/keda_hpa.go:98","message":"Creating Scaled Object \"metrics-test-00001\"","commit":"6f54184-dirty","knative.dev/controller":"knative.dev.autoscaler-keda.pkg.reconciler.autoscaling.hpa.Reconciler","knative.dev/kind":"autoscaling.internal.knative.dev.PodAutoscaler","knative.dev/traceid":"0fbaf41d-ecbf-4cf5-b417-64b2543e2e7f","knative.dev/key":"default/metrics-test-00001"}
```

Scaling of only one service:
<img width="748" alt="Screenshot 2024-11-19 at 11 26 47 PM" src="https://github.com/user-attachments/assets/6e6c4923-abb4-4624-8099-32698bda119d">

<!--
In addition, categorize the changes you're making using the "/kind" Prow command, example:

/kind <kind>

Supported kinds are: api-change, bug, cleanup, deprecation, removal, documentation, enhancement, performance

-->
/kind documentation
/kind enhancement

skonto commented on pkg/reconciler/autoscaling/hpa/resources/keda.go:
```diff
@@ -128,6 +131,20 @@ func DesiredScaledObject(ctx context.Context, pa *autoscalingv1alpha1.PodAutosca
 				return nil, fmt.Errorf("query is missing for custom metric: %w", err)
 			}
 
+			tmpl, err := template.New("query").Funcs(sprig.TxtFuncMap()).Parse(query)
```
Sprig has more functions but for now let's keep this minimal, just use the Golang templates.

muscionig replied:
I removed `sprig`.

skonto commented on go.mod:
```diff
@@ -23,6 +23,10 @@ require (
+	dario.cat/mergo v1.0.1 // indirect
+	github.com/Masterminds/goutils v1.1.1 // indirect
+	github.com/Masterminds/semver/v3 v3.3.0 // indirect
+	github.com/Masterminds/sprig/v3 v3.3.0 // indirect
```
Updating go.mod required to run ./hack/update-deps.sh

muscionig replied:
Thanks for this! I noticed something was not right when I was running my typical update, should be fixed now.

skonto commented on pkg/reconciler/autoscaling/hpa/resources/keda.go:
```diff
@@ -128,6 +131,20 @@ func DesiredScaledObject(ctx context.Context, pa *autoscalingv1alpha1.PodAutosca
+				return nil, fmt.Errorf("template initialization failed: %w", err)
+			}
+			values := map[string]string{
+				"revisionName": pa.Name,
```
As a side note Serving supports byo revision names instead of being auto-generated. 
```
kind: Service
metadata:
  name: hello
spec:
  template:
    metadata:
      name: hello-1
```
If you want to change that service then you need to manually change `hello-1` to something else. PA name is always the same as with revision name.

muscionig replied:
Thanks! I think it would be helpful to include this, along with the other points we've discussed, in a "Usage Recommendations" document. Should we log an issue to capture these notes? That way, we can reference it and update it with new considerations as they come up.

muscionig replied:
issue #86

skonto commented:
@muscionig let's proceed with this and add some tests to verify that it works as expected.

muscionig commented:
@skonto Thanks for the review! I've addressed the dependencies and added a test case. I have a couple of questions:

1. Would you add logic to validate user input? For example, if the query annotation contains anything other than `revisionName` (e.g., `sum(rate(http_requests_total{pod=~\"{{ .wrongValue }}.*\"}[1m]))`), the output becomes `sum(rate(http_requests_total{pod=~"<no value>.*"}[1m]))`.
2. Should we gate this feature through the configMap? Personally, I believe it should always be enabled.

skonto commented:
Hi @muscionig 
1. Yes we should validate the input.
2. This should be enabled by default and we should allow to disable it on demand per service (just in case some use case needs it).

muscionig commented:
@skonto I worked on the `e2e` tests and added two:

1. **`TestUpDownCustomMetricTemplate`**: This test checks that a service with the templated Prometheus query can be created.
2. **`TestMultiRevisionScalingCM`**: A more comprehensive test involving a service with two revisions. It verifies that:
   - Only the second revision scales up.
   - The first revision remains unscaled.

**Notes:**
      1. I have tried to reuse as much of the existing definitions as possible, hence, I kept the revisions in two separate `TestContext`s. Please let me know if this approach is acceptable.
      2. Since the `Revision` is hardcoded in `helpers.go`, I created a new function to get the second revision. Given the size of this PR, I avoided refactoring the `GetResourceObject` functions, which seemed excessive at this stage.


On the additions we mentioned. A couple of questions for clarification:
1. **Input validation**: I’m planning to validate that if a template is present (via regex), it can only contain `.revisionName`. I looked at `go/template` and, unless I missed something, it doesn’t natively support this kind of validation.
2. **Disabling annotation**: I'll add a `disable-template-substitution` annotation and include checks for it.

The `e2e` tests took quite a bit of effort to figure out, which is why I haven’t finished the additional features yet. I’d love to hear your feedback on those tests and let me know whether the 2 additions should be handled in separate PRs.

skonto commented on test/e2e/autoscale_custom_test.go:
```diff
@@ -76,6 +98,26 @@ func TestUpDownCustomMetric(t *testing.T) {
 	assertCustomHPAAutoscaleUpToNumPods(ctx, targetPods, time.After(scaleUpTimeout), true /* quick */)
 }
 
+func TestUpDownCustomMetricTemplate(t *testing.T) {
```
I think this is redundant as we have the multi revision test where we test the template feature.

muscionig replied:
removed

skonto commented on test/e2e/autoscale_custom_test.go:
```diff
@@ -57,6 +57,28 @@ const (
 	targetPods            = 2
 )
 
+func TestMultiRevisionScalingCM(t *testing.T) {
```
Works as expected:

```
  autoscale.go:223: revision "multi-revision-scaling-c-m-tpptghkr-00002" #replicas: 2, want at least: 1
    autoscale.go:231: Quick Mode: got 2 >= 2
    autoscale.go:119: Stopping generateTraffic
    stream.go:304: I 12:01:56.257 autoscaler-6c7bf97997-mx876 [knative.dev.serving.pkg.reconciler.autoscaling.kpa.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Reconcile succeeded
    stream.go:304: I 12:01:56.280 activator-d66fd5dd8-ctz76 [activator] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Updating Revision Throttler with: clusterIP = <nil>, trackers = 2, backends = 2
    stream.go:304: I 12:01:56.280 activator-d66fd5dd8-ctz76 [activator] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Set capacity to 2147483647 (backends: 2, index: -1/0)
    stream.go:304: I 12:01:56.280 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.serverlessservice.reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] SKS is in Serve mode; has 1 endpoints in multi-revision-scaling-c-m-tpptghkr-00002-private; 1 activator endpoints
    stream.go:304: I 12:01:56.280 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.serverlessservice.reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Public K8s Endpoints changed; reconciling: multi-revision-scaling-c-m-tpptghkr-00002
    stream.go:304: I 12:01:56.288 autoscaler-6c7bf97997-mx876 [knative.dev.serving.pkg.reconciler.autoscaling.kpa.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Reconcile succeeded
    stream.go:304: I 12:01:56.295 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.serverlessservice.reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Reconcile succeeded
    stream.go:304: I 12:01:56.299 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.serverlessservice.reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] SKS is in Serve mode; has 2 endpoints in multi-revision-scaling-c-m-tpptghkr-00002-private; 1 activator endpoints
    stream.go:304: I 12:01:56.299 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.serverlessservice.reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Public K8s Endpoints changed; reconciling: multi-revision-scaling-c-m-tpptghkr-00002
    stream.go:304: I 12:01:56.307 net-kourier-controller-5db85876d8-t5bn7 [knative.dev.net-kourier.pkg.reconciler.ingress.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr] Updating Ingress
    stream.go:304: I 12:01:56.307 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.serverlessservice.reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Reconcile succeeded
    stream.go:304: I 12:01:56.307 net-kourier-controller-5db85876d8-t5bn7 [knative.dev.net-kourier.pkg.reconciler.ingress.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr] Reconcile succeeded
    stream.go:304: I 12:01:56.587 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.revision.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Reconciling PA: multi-revision-scaling-c-m-tpptghkr-00002
    stream.go:304: I 12:01:56.587 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.revision.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr-00002] Reconcile succeeded
    autoscale.go:223: revision "multi-revision-scaling-c-m-tpptghkr-00001" #replicas: 0, want at least: 0
    autoscale.go:231: Quick Mode: got 0 >= 0
    stream.go:304: I 12:01:56.821 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.service.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr] Reconcile succeeded
    stream.go:304: I 12:01:56.835 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.service.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr] Reconcile succeeded
    stream.go:304: I 12:01:56.835 controller-5b54cd98c-8td46 [knative.dev.serving.pkg.reconciler.route.Reconciler] [serving-tests/multi-revision-scaling-c-m-tpptghkr] Reconcile succeeded
--- PASS: TestMultiRevisionScalingCM (128.36s)
PASS
ok  	knative.dev/autoscaler-keda/test/e2e	128.383s
````

skonto commented on test/e2e/autoscale_custom_test.go:
```diff
@@ -57,6 +57,28 @@ const (
+	assertCustomHPAAutoscaleUpToNumPods(ctxRevision2, targetPods, time.After(scaleUpTimeout), true /* quick */)
+	// at completion of the previous, the revision in ctxRevision1 should be scaled to 0
+	assertAutoscaleUpToNumPods(ctxRevision1, 0, time.After(scaleUpTimeout), true /* quick */, 0)
+	// TODO: in addition we could create an additional assert that explicitly checks that the revision that should not be scaled up is not scaled up
```
Could you add this part?

muscionig replied:
done

skonto commented:
@muscionig let's turn this to a regular PR, address the comments above and then test.

muscionig commented:
@skonto I added this in my latest commit. I did not defined additional functions, just reused what was there and added a loop to check the scale up status of `ctxRevision1`. 

PR is ready for review.

skonto commented:
/ok-to-test

muscionig commented on pkg/reconciler/autoscaling/hpa/resources/keda.go:
```diff
@@ -128,6 +130,20 @@ func DesiredScaledObject(ctx context.Context, pa *autoscalingv1alpha1.PodAutosca
+			if err := tmpl.Execute(&output, values); err != nil {
+				return nil, fmt.Errorf("template execution failed: %w", err)
+			}
+			log.Printf("PrometheusQuery rendered: %s\n", output.String())
```
I have removed it for now. Wanna log an issue for it?

skonto commented:
/approve

skonto commented:
/lgtm
